### ISSUE TITLE
Decrypted result returned false

### ISSUE THREAD

**Issue by Xdudu** [travist/jsencrypt#103]
So appreciated that I found this library! But when I use it in my project, the decrypted result always returns false. Does anyone know what possibly happened? In which cases decrypt.decrypt(encrypted_info) may return false? Thanks a lot!!

**Comment by h3dx0** [comment]
Any solution for this problem @Xdudu ?

**Comment by alfaro28** [comment]
It may be because your key has an unsupported format, as it only supports 
```
-----BEGIN PUBLIC KEY-----
xxx...
-----END PUBLIC KEY-----
```
machinewu added support for 
```
-----BEGIN RSA PUBLIC KEY-----
xxx...
-----END RSA PUBLIC KEY-----
```
so you can use his fork or mine (I forked from him) to work with those keys

**Comment by GingerSnap-xx** [comment]
I recently had a problem with this crop up after having successfully used the library. For some reason it stopped overwriting the key if you created added it by setting they key instead of passing it into the JSEncrypt constructor. If you aren't passing in the RSA key to the constructor try that.

**Comment by seungho** [comment]
I got the same error.
I used jsencrypt.js instead of jsencrypt.min.js and it worked.
Thanks!

**Comment by gramo44** [comment]
I got the error just when the message to encrypt is long. Supported lenght depends on the bitlenght of the key
i.e.
If I use private_key_bits of 1024 I can send
`"José compró en Perú una vieja zampoña. Excusándose, Sofía tiró su whisky al desagüe de la banqueta."
`nothing longer, after that jsencrypt returns false.

If I use private_key_bits of 512 I can send
`"José compró en Perú una vieja zampoña. Excusánd"
`nothing longer, after that jsencrypt returns false.

On long strings JavaScript console reports: "Message too long for RSA"

My working code in php is as follows:

```
<?php
    //------------------------------------------------------------
    // Global Settings.
    //------------------------------------------------------------
    ini_set('display_errors', 1);
    error_reporting(E_ALL);
    $directorio = "/path/to/key/directory/apache/writable/";
    $nombre_base = "llaves_php";

    //------------------------------------------------------------
    // Initialization.
    //------------------------------------------------------------
    $encabezado_html = "";
    $cuerpo_html = "";

    //------------------------------------------------------------
    // Loading keys
    //------------------------------------------------------------
    list($privateKey, $pubKey) =
        cargar_llaves_RSA($directorio, $nombre_base);

    //------------------------------------------------------------
    // Form that uses javascript to encrypt data.
    // (it uses only the public key)
    //------------------------------------------------------------
    $librerias_html = "
        <script type='text/javascript'
                src='https://ajax.googleapis.com/ajax/libs/".
                    "jquery/3.2.1/jquery.min.js'></script>
        <script type='text/javascript'
                src='lib/jsencrypt.js'></script>
        ";

    $pubKey_html = htmlentities($pubKey);
    $datos_html = "
        <h2>Cifrando con Javascript</h2>
        <input type='text' id='mensaje' />
        <br />
        <button id='ENVIAR'>Enviar</button>
        <br />
        <textarea id='pubkey' style='display: none;'>".
        $pubKey_html.
        "</textarea>
        <script type='text/javascript'>
            $('#ENVIAR').click(function () {
                var codificador = new JSEncrypt();
                codificador.setKey($('#pubkey').val());
                var cifrado = codificador.encrypt($('#mensaje').val());
                window.open('?mensaje=' + encodeURIComponent(cifrado)
                           , '_top');
            });
        </script>
        ";

    //------------------------------------------------------------
    // Decrypting using php (it uses only the privateKey)
    //------------------------------------------------------------
    if (isset($_REQUEST['mensaje'])) {
        openssl_private_decrypt( base64_decode($_REQUEST['mensaje'])
                               , $descifrado
                               , $privateKey);
        $datos_html.= "
            <h2>Descifrando con PHP</h2>
            ".$descifrado."
            ";
    }

    //------------------------------------------------------------
    // HTML DISPLAY
    //------------------------------------------------------------
    $encabezado_html.= "<title>Receptor de mensaje cifrado</title>"
                     . $librerias_html;

    $cuerpo_html.= $datos_html;

    $contenido = "<head>$encabezado_html</head><body>$cuerpo_html</body>";
    $contenido = "<html>$contenido</html>";
    print $contenido;
    
//============================================================
//============================================================
// Functions
//============================================================
//============================================================

    //------------------------------------------------------------
    function cargar_llaves_RSA($directorio, $nombre_base) {
    //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    // PROPÓSITO: Genera o carga desde archivos las llaves RSA
    // ENTRADAS:
    // $directorio: Directorio donde se encuentran los archivos.
    // $nombre_base: Nombre, sin extensión, de los archivos con
    //               las llaves.
    // SALIDAS:
    //------------------------------------------------------------
        if (  !file_exists($directorio.$nombre_base.".crt")
           || !file_exists($directorio.$nombre_base.".pub")) {
            list($privateKey, $pubKey) =
                crear_llaves_RSA($directorio.$nombre_base);
        } else {
            //------------------------------------------------------------
            // CARGA DE LLAVES RSA ARCHIVADAS
            //------------------------------------------------------------
            $privateKey = file_get_contents($directorio.$nombre_base.".crt");
            if (!$privKey = openssl_pkey_get_private($privateKey))
                die('Loading Private Key failed');
            $pubKey  = file_get_contents($directorio.$nombre_base.".pub");
        }

        return array($privateKey, $pubKey);
    }

    //------------------------------------------------------------
    function crear_llaves_RSA($ruta_base) {
    //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    // PROPÓSITO:
    // generacion de llaves RSA en php
    // ENTRADAS:
    // $ruta_base: Ruta de los archivos a generar sin extensión.
    // SALIDAS:
    // Se generarán dos archivos, uno con la llave privada con
    // extensión .crt, el otro con llave pública con extensión
    // .pub; la función retorna tanto la llave pública como la
    // privada en un arreglo.
    //------------------------------------------------------------
        $config = array(
            "private_key_bits" => 1024,
            "private_key_type" => OPENSSL_KEYTYPE_RSA,
        );

        $llavePrivadaCruda = openssl_pkey_new($config);
        openssl_pkey_export_to_file($llavePrivadaCruda, $ruta_base.".crt");
        $privateKey = file_get_contents($ruta_base.".crt");
        openssl_pkey_export($llavePrivadaCruda, $privKey);

        $pubKeyData = openssl_pkey_get_details($llavePrivadaCruda);
        $pubKey = $pubKeyData["key"];
        file_put_contents($ruta_base.".pub", $pubKey);
        openssl_free_key($llavePrivadaCruda);

        return array($privateKey, $pubKey);
    }

    //------------------------------------------------------------
    function Mostrar($valor) {
    //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    // PROPÓSITO: Genera el código HTML para presentar una
    // variable embebida en la página.
    // ENTRADAS:
    // $valor: el valor a presentar.
    // SALIDAS: código html que permite visualizar la variable.
    //------------------------------------------------------------
        $retorno = htmlentities(stripslashes(var_export($valor, true)));
        $retorno = "<pre>$retorno</pre>";
        return $retorno;
    }

?>
```
Directory tree must looks like:
```
├── script.php
└── lib
    └── jsencrypt.js
```
and a directory writable by php outside of public zone named
`/path/to/key/directory/apache/writable/`

**Comment by gramo44** [comment]
I think I solve the question about string lengths on:
https://security.stackexchange.com/questions/33434/rsa-maximum-bytes-to-encrypt-comparison-to-aes-in-terms-of-security/33445#33445

> RSA, as defined by PKCS#1, encrypts "messages" of limited size. With the commonly used "v1.5 padding" and a 2048-bit RSA key, the maximum size of data which can be encrypted with RSA is 245 bytes. No more.

**Comment by danthe1st** [comment]
I have the same problem, but I use Java Servlets.
maybe the generated Key from PHP/JServlet is not valid for js.

**Comment by flocosta19** [comment]
Hi all,

openssl genrsa -out rsa_2048_priv.pem 2048
openssl rsa -pubout -in rsa_1024_priv.pem -out rsa_1024_pub.pem

The key pairs generation above does not use -des3 type so it does not prompt for the passphrase.

openssl genrsa -des3 -out rsa_1024_priv.pem 1024
openssl rsa -pubout -in rsa_1024_priv.pem -out rsa_1024_pub.pem
This will prompt passphrase.

Now, the JSEncrypt library does not support with passphrase prompt.

Hope this helps.

**Comment by danthe1st** [comment]
Maybe an invalid Key has been used, for example you did not use Base64 encryption but something different.
